tests/test_mock_api.py

```diff
@@ -0,0 +1,288 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Tests for the microgrid mock api."""
+
+# pylint: disable=missing-function-docstring,use-implicit-booleaness-not-comparison
+# pylint: disable=invalid-name,no-name-in-module,no-member
+
+from unittest.mock import Mock
+
+import grpc
+from frequenz.api.common.components_pb2 import ComponentCategory
+from frequenz.api.microgrid.microgrid_pb2 import (
+    Component,
+    ComponentFilter,
+    Connection,
+    ConnectionFilter,
+)
+from frequenz.api.microgrid.microgrid_pb2_grpc import MicrogridStub
+from google.protobuf.empty_pb2 import Empty
+
+from . import mock_api
+
+
+def test_MockMicrogridServicer() -> None:
+    """Test the MockMicrogridServicer."""
+    api = mock_api.MockMicrogridServicer()
+    service_context_mock = Mock(spec=grpc.ServicerContext)
+    assert (
+        list(api.ListComponents(ComponentFilter(), service_context_mock).components)
+        == []
+    )
+    assert (
+        list(api.ListConnections(ConnectionFilter(), service_context_mock).connections)
+        == []
+    )
+
+    # adding new components just appends them to the list:
+    # duplicates are not prevented, as this may be wanted
+    # behaviour of the mock
+    api.add_component(0, ComponentCategory.COMPONENT_CATEGORY_METER)
+    assert list(
+        api.ListComponents(ComponentFilter(), service_context_mock).components
+    ) == [Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_METER)]
+    assert (
+        list(api.ListConnections(ConnectionFilter(), service_context_mock).connections)
+        == []
+    )
+
+    api.add_component(0, ComponentCategory.COMPONENT_CATEGORY_BATTERY)
+    assert list(
+        api.ListComponents(ComponentFilter(), service_context_mock).components
+    ) == [
+        Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_METER),
+        Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_BATTERY),
+    ]
+    assert (
+        list(api.ListConnections(ConnectionFilter(), service_context_mock).connections)
+        == []
+    )
+
+    api.add_component(0, ComponentCategory.COMPONENT_CATEGORY_METER)
+    assert list(
+        api.ListComponents(ComponentFilter(), service_context_mock).components
+    ) == [
+        Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_METER),
+        Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_BATTERY),
+        Component(id=0, category=ComponentCategory.COMPONENT_CATEGORY_METER),
+    ]
+    assert (
+        list(api.ListConnections(ConnectionFilter(), service_context_mock).connections)
+        == []
+    )
+
+    # similarly, duplicates are allowed when adding new connections
+    api.add_connection(0, 0)
```

Comment:
I guess `(0, 0)` refer to the source and destination nodes? If yes, then there is a cycle in the component DAG, which makes it not-a-DAG. Why is this not an error case?

Reply:
:cd: I will also defer this question to @shsms and wait for the answer to see if we need a separate issue for this.

Reply:
agree here too, and it is from before my time on the SDK.